src/util/enumerate.rs

```diff
@@ -0,0 +1,80 @@
+use futures::{Async, Poll, Stream, Sink, StartSend};
+
+/// A stream combinator which combines the yields the current item
+/// plus its count starting from 0.
+///
+/// This structure is produced by the `Stream::enumerate` method.
+#[derive(Debug)]
+#[must_use = "Does nothing unless polled"]
+pub struct Enumerate<T> {
+    inner: T,
+    count: usize,
+}
+
+impl<T> Enumerate<T> {
+    pub fn new(stream: T) -> Self {
```

Comment:
Can we avoid making this public? Construction can happen via the stream trait.

Reply:
Sorry, not entirely sure how that works. Can you give me a clue. Essentially we need to make it module private but how does that tie in with the pub export that you mentioned earlier ?

Reply:
`pub(crate) fn new(...)` then call `Enumerate::new` in `StreamExt::enumerate().